src/core/async_call.js

```diff
@@ -0,0 +1,75 @@
+mergeInto(LibraryManager.library, {
+  temp: function () {
+    if (typeof WebAssembly.Function === "undefined") {
+      return;
+    }
+    function sleep(ms) {
+      return new Promise((res) => setTimeout(res, ms));
+    }
+    const _orig_Py_CheckEmscriptenSignals_Helper =
+      _Py_CheckEmscriptenSignals_Helper;
+    const wrapper = async function () {
+      await sleep(0);
+      return _orig_Py_CheckEmscriptenSignals_Helper();
+    };
+    const suspending__Py_CheckEmscriptenSignals_Helper =
+      new WebAssembly.Function(
+        { parameters: ["externref"], results: ["i32"] },
+        wrapper,
+        { suspending: "first" },
+      );
+    const bytes = [
+      0, 97, 115, 109, 1, 0, 0, 0, 1, 10, 2, 96, 1, 111, 1, 127, 96, 0, 1, 127,
```

Comment:
What's this?

Reply:
The wat currently looks like:
<details>
<summary>Syncify wat</summary>


```wat
(module
    (global $suspender (import "e" "s") (mut externref))
    ;; Status flag to tell us if suspender is usable
    (global $check (import "e" "c") (mut i32))
    ;; Wrapped syncify function. Expects suspender as a first argument and a
    ;; JsRef to promise as second argument. Returns a JsRef to the result.
    (import "e" "i" (func $syncify_promise_import (param externref i32) (result i32)))
    ;; Wrapped syncify_promise that handles suspender stuff automatically so
    ;; callee doesn't need to worry about it.
    (func $syncify_promise_export (export "o")
      (param $idpromise i32) (result i32)
      (local $suspender_copy externref)
      (global.get $check)
      (i32.eqz)
      if
        ;; If no valid suspender, return 0. Callee needs to check for this case
        ;; and set Python error flag (or otherwise handle it).
        i32.const 0
        return
      end
      ;; Store suspender into a local variable in order to manage reentrancy
      (global.get $suspender)
      (local.tee $suspender_copy)
      (local.get $idpromise)
      (call $syncify_promise_import) ;; onwards call args are (suspender, orig argument)
      (local.get $suspender_copy)
      (global.set $suspender) ;; restore global variable
    )
)
```



</details>
The check signals wat is similar:

<details>
<summary>check signals wat</summary>

```wat
(module
    (global $suspender (import "e" "s") (mut externref))
    (global $check (import "e" "c") (mut i32))
    (import "e" "i" (func $CheckSignals (result i32)))
    (import "e" "r" (func $await_sleep_0 (param externref)))
    (func $suspending_call_export (export "o") (result i32)
      (local $suspender_copy externref)
      (global.get $check)
      if
        ;; If we have a valid suspender, insert an await sleep 0 to unblock
        ;; event loop.
        (global.get $suspender)
        (local.tee $suspender_copy)
        (call $await_sleep_0)
        (local.get $suspender_copy)
        (global.set $suspender)
      end
      ;; In any case call original CheckSignals function
      (call $CheckSignals)
    )
)
```

</details>

The `__pyproxy_apply` wrapper receives a suspender as the first argument which it stores into the suspender global variable. Then it plugs the rest of the arguments into the wrapped function:
<details>
<summary>__pyproxy_apply wat</summary>

```wat
(module
    (global $suspender (import "e" "s") (mut externref))
    (import "e" "i" (func $apply_import (param i32 i32 i32 i32 i32) (result i32)))
    (func $apply_export (export "o")
      (param $susp externref) (param $a i32) (param $b i32) (param $c i32) (param $d i32) (param $e i32) (result i32)
      ;; Store first variable into suspender global
      (local.get $susp)
      (global.set $suspender)
      ;; Plug remaining args into onwards call
      (local.get $a)
      (local.get $b)
      (local.get $c)
      (local.get $d)
      (local.get $e)
      (call $apply_import)
    )
)
```
</details>